File: packages/aurora/src/main/java/com/sofar/aurora/feature/home/HomeFragment.java
package com.sofar.aurora.feature.home;

import com.sofar.aurora.R;
import com.sofar.aurora.feature.home.model.HomeBlock;
import com.sofar.aurora.feature.home.model.HomePageList;
import com.sofar.base.page.PageList;
import com.sofar.base.recycler.RecyclerAdapter;
import com.sofar.base.recycler.RecyclerFragment;

public class HomeFragment extends RecyclerFragment<HomeBlock> {

  @Override
  protected RecyclerAdapter<HomeBlock> onCreateAdapter() {
    return new HomeAdapter();
  }

  @Override
  protected PageList<?, HomeBlock> onCreatePageList() {
    return new HomePageList();
  }

  @Override
  protected int getLayoutResId() {
    return R.layout.home_fragment;
  }
}

File: packages/wan-android/src/main/java/com/sofar/wan/android/feature/home/HomeFragment.kt
package com.sofar.wan.android.feature.home

import android.os.Bundle
import android.view.View
import androidx.recyclerview.widget.DiffUtil
import com.sofar.wan.android.feature.article.ArticleCell
import com.sofar.wan.android.feature.article.ArticleDiffCalculator
import com.sofar.wan.android.feature.article.ArticleUtil
import com.sofar.wan.android.feature.banner.BannerCell
import com.sofar.wan.android.model.Article
import com.sofar.wan.android.model.Banners
import com.sofar.wan.android.paging.PageFragment
import com.sofar.wan.android.paging.PageList
import com.sofar.widget.recycler.adapter.CellAdapter
import com.sofar.widget.recycler.adapter.multitype.MultiTypeAdapter

class HomeFragment : PageFragment<Any>() {

  override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
    super.onViewCreated(view, savedInstanceState)
    recyclerView.addItemDecoration(ArticleUtil.createItemDecoration())
  }

  override fun onCreateAdapter(): CellAdapter<Any> {
    var adapter = MultiTypeAdapter()
    adapter.register(Article::class.java) { ArticleCell() }
    adapter.register(Banners::class.java) { BannerCell() }
    return adapter
  }

  override fun onCreatePageList(): PageList<*, Any> {
    return HomePageList()
  }

  override fun onCreateDiffCallback(): DiffUtil.ItemCallback<Any> {
    return ArticleDiffCalculator.getCommonDiffItemCallback()
  }
}